dftimewolf/lib/containers/manager.py

```diff
@@ -172,28 +186,49 @@ def CompleteModule(self, module_name: str) -> None:
       raise RuntimeError("Container manager has not parsed a recipe yet")
 
     with self._mutex:
-      self._modules[module_name].completed = True
-
-      # With that `module_name` marked completed, any modules with all
-      # dependants completed can have storage cleared.
-      for name, module in self._modules.items():
-        if self._CheckDependenciesCompletion(name):
-          for c in module.storage:
-            del c
-          module.storage = []
-
-  def _CheckDependenciesCompletion(self, module_name: str) -> bool:
-    """For a module, checks if other modules that depend on are complete.
-
+      self._modules[module_name].storage = {}
+
+  def RegisterStreamingCallback(
+      self,
+      module_name: str,
+      container_type: Type[T],
+      callback: Callable[[interface.AttributeContainer], None]) -> None:
+    """Registers a container streaming callback for a module and container type.
+    
     Args:
-      module_name: The module name to check for.
-
-    Returns:
-      True if all modules dependant on the named module have complete; False
-          otherwise.
+      module_name: The module name registering the callback
+      container_type: The container type to stream
+      callback: The function to call with containers
     """
-    for _, module in self._modules.items():
-      if module_name in module.dependencies:
-        if not module.completed:
-          return False
-    return True
+    if not self._modules:
+      raise RuntimeError('Container manager has not parsed a recipe yet')
+    if module_name not in self._modules:
+      raise RuntimeError('Registering a callback for a non-existent module')
+
+    if container_type.CONTAINER_TYPE not in self._modules[module_name].callback_map:
+      self._modules[module_name].callback_map[container_type.CONTAINER_TYPE] = []
```

Comment:
Isn't callback_map already a defaultdict?

Reply:
Added methods to the helper class.